strutil/set.go

```diff
@@ -0,0 +1,96 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+// OrderedSet is a set of strings that maintains the order of insertion.
+//
+// External synchronization is required for safe concurrent access.
+type OrderedSet struct {
+	positionOf map[string]int
+}
+
+// Items returns a slice of strings representing insertion order.
+//
+// Contains is O(N) in the size of the set.
+func (o *OrderedSet) Items() []string {
+	if o.positionOf == nil {
+		return nil
+	}
+	items := make([]string, len(o.positionOf))
+	for item, idx := range o.positionOf {
+		items[idx] = item
+	}
+	return items
+}
+
+// Contains returns true if the set contains a given item.
+//
+// Contains is O(1) in the size of the set.
+func (o *OrderedSet) Contains(item string) bool {
+	if o.positionOf == nil {
+		return false
+	}
+
+	_, ok := o.positionOf[item]
+	return ok
+}
+
+// IndexOf returns the position of an item in the set.
+func (o *OrderedSet) IndexOf(item string) (idx int, ok bool) {
+	idx, ok = o.positionOf[item]
+	return idx, ok
+}
+
+// Del removes an item from the set.
+//
+// Delete is O(1) in the size of the set.
+func (o *OrderedSet) Del(item string) {
+	if o.positionOf == nil {
+		return
+	}
+
+	delete(o.positionOf, item)
+	if len(o.positionOf) == 0 {
+		o.positionOf = nil
+	}
+}
+
+// Put adds an item into the set.
+//
+// If the item was not present then it is stored and ordered after all existing
+// elements. If the item was already present its position is not changed.
+//
+// Put is O(1) in the size of the set.
+func (o *OrderedSet) Put(item string) {
+	if o.positionOf == nil {
+		o.positionOf = make(map[string]int)
+	}
+
+	if _, ok := o.positionOf[item]; ok {
+		return
+	}
+
+	o.positionOf[item] = len(o.positionOf)
```

Comment:
while potentially clever, I think this is wrong, see https://play.golang.org/p/Pp6vUyS3Uzp specifically, but what happens when we delete enough things that the length goes below something inserted earlier, then insert another thing? the newly inserted thing will be ordered _before_ the old thing and thus the ordering is broken. see:

```go
	s := OrderedSet{}
	s.Put("a") // a is 0
	fmt.Printf("%+v\n", s.positionOf)
	s.Put("b") // b is 1
	fmt.Printf("%+v\n", s.positionOf)
	s.Put("c") // c is 2
	fmt.Printf("%+v\n", s.positionOf)
	s.Del("a")
	fmt.Printf("%+v\n", s.positionOf)
	s.Del("b")
	fmt.Printf("%+v\n", s.positionOf)
	s.Put("d") // d is now len() == 1, but this is less than "c" which is 2
	fmt.Printf("%+v\n", s.positionOf)
```

I think instead an easy patch for this is to just keep a running counter, and increment it for every `Put`, but never decrement it.

Reply:
nice

Reply:
Silly me, I'll have a look at fixing this now.

Reply:
We don't need to delete elements in the initial use case so I'll just skip implementing delete altogether. Since the implementation is private we can change it at any time, if we end up needing it in another spot.

Thank you for spitting this again.

Reply:
sure, if we don't need to delete things then the current code is fine